File: nnp1.py
import numpy as np

inputs = [[1,2,3,2.5], [2.0, 5.0, -1.0, 2.0], [-1.5, 2.7, 3.3, -0.8]]

weights1 = [0.2, .8, -.5, 1]
weights2 = [.5, -.91, .26, -0.5]
weights3 = [-.26, -.27, .17, .87]

weights = [weights1,weights2, weights3]
biases = [2, 3, .5]

#layer_outputs = [] #output of current layer
#zip combines two lists by pairing the nth element of the first list with the nth element second list
#e.g [[weights[0], biases[0]], [weights[0], biases[0]]] or [[[0.2, .8, -.5, 1], 2], [[.5, -.91, .26, -0.5], 3]]
#the following for in loop will iterate through each element of the zipped tuple
# for neuron_weights, neuron_bias in zip(weights, biases):
#   neuron_output = 0 # Output of given neuron
#   for n_input, weight in zip(inputs, neuron_weights):
#     neuron_output += n_input * weight
#   neuron_output += neuron_bias
#   layer_outputs.append(neuron_output)
  
neuron_output = np.dot(inputs, np.array(weights).T) + biases
print(neuron_output)
# print([2]+[5])
# print(layer_outputs)

File: nnp1.js
const inputs = [1, 2, 3, 2.5];

const weights1 = [0.2, 0.8, -0.5, 1];
const weights2 = [0.5, -0.91, 0.26, -0.5];
const weights3 = [-0.26, -0.27, 0.17, 0.87];
const weights = [weights1, weights2, weights3];

const biases = [2, 3, 0.5];

const layerOutput = [];

for (let [neuronWeights, neuronBias] of zip(weights, biases)) {
  let neuronOutput = 0;
  for (let [neuronWeight, neuronInput] of zip(neuronWeights, inputs)) {
    neuronOutput += neuronInput * neuronWeight;
  }
  neuronOutput += neuronBias;
  layerOutput.push(neuronOutput);
}

console.log(layerOutput);
console.log(dot(weights, inputs));

function zip(list1, list2) {
  let newList = [];
  for (let i = 0; i < list1.length; i++) {
    if (list1[i] !== undefined && list2[i] !== undefined) {
      newList.push([list1[i], list2[i]]);
    }
  }
  return newList;
}

function dot(vector1, vector2) {
  let returnVector = [];
  if (vector1[0][0] !== undefined) {
    vector1.forEach((vector) => returnVector.push(dot(vector, vector2)));
  }
  if (returnVector.length === 0) returnVector = vector1;
  else return returnVector;
  if (returnVector.length !== vector2.length) throw new Error("Invalid shape");
  returnVector = returnVector
    .map((val, i) => val * vector2[i])
    .reduce((reduced, curr) => reduced + curr, 0);
  return returnVector.length <= 1 ? returnVector[0] : returnVector;
}
